### PR TITLE
Use the gpu-to-llvm-spv pass to lower some gpu operations

### PR DESCRIPTION
**Pull request by FMarno**
These changes will remove the scalar id operations from TritonGEN and will instead lower the gpu operations directly to llvm via the `gpu-to-llvm-spv-pass`.

There will be future PRs to remove more of the TritonGEN dialect as the lowering is ready in upstream LLVM.

### CODE REVIEW

File: scripts/test-triton.sh [hunk 1]
```diff
@@ -162,10 +162,10 @@ run_core_tests() {
   export TEST_UNSKIP
 
   TRITON_DISABLE_LINE_INFO=1 TRITON_TEST_SUITE=language \
-  pytest -vvv -n 8 --device xpu language/ --ignore=language/test_line_info.py --ignore=language/test_subprocess.py
+  pytest -vvv -n 32 --device xpu language/ --ignore=language/test_line_info.py --ignore=language/test_subprocess.py
```
Comment (FMarno) [inline scripts/test-triton.sh:165]: revert this change

File: third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp [hunk 2]
```diff
@@ -128,7 +128,7 @@ class CallOpConversion : public mlir::RewritePattern {
     auto operands = callOp.getOperands();
     auto result = callOp.getResult();
 
-    LLVM::LLVMFunctionType calleeType = callOp.getCalleeType().value();
+    LLVM::LLVMFunctionType calleeType = callOp.getVarCalleeType().value();
```
Comment (victor-eds) [inline third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp:131]: Might be wrong, but this looks like it's querying indirect calls call types?

**Review by victor-eds — APPROVED** [review]

File: third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp (hunk 2 above)
Comment (FMarno) [reply third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp:131]: This is nothing to do with the gpu to llvm spv pass, someone changed the name of `$callee` to `$var_callee` upstream.

Comment (victor-eds) [reply third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp:131]: Yeah, just curiosity. Code LGTM.

**Comment by whitneywhtsang** [comment]
FYI @FMarno https://github.com/intel/intel-xpu-backend-for-triton/pull/1940 upgraded LLVM commit.

**Comment by FMarno** [comment]
> FYI @FMarno #1940 upgraded LLVM commit.

Unfortunately that update actually only moved the LLVM forwards about 1.5 hours :disappointed:.

**Comment by FMarno** [comment]
> FYI @FMarno #1940 upgraded LLVM commit.

Just for clarity, the current llvm commit has is for the 28th of June, but the changes I need were added on the 24th of July.

**Comment by whitneywhtsang** [comment]
FYI @FMarno https://github.com/intel/intel-xpu-backend-for-triton/pull/2075 upgraded LLVM commit.

**Comment by FMarno** [comment]
@whitneywhtsang Thanks, that should be enough for the first set of operations.

**Comment by FMarno** [comment]
@whitneywhtsang @etiotto This is ready for review now.

File: third_party/intel/CMakeLists.txt [hunk 3]
```diff
@@ -14,6 +14,7 @@ add_triton_plugin(TritonXPU
   TritonIntelGPUToLLVM
   TritonIntelGPUTransforms
   TritonToTritonGPUWarp
+  MLIRGPUToLLVMSPV
```
Comment (whitneywhtsang) [inline third_party/intel/CMakeLists.txt:17]: alphabetical order?

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h [hunk 4]
```diff
@@ -262,6 +263,8 @@ class TritonGPUToLLVMPipelineManager {
     triton::populateTritonGENToLLVMConversionPatterns(typeConverter, patterns);
     triton::populateGPUToTritonGENConversionPatterns(typeConverter, patterns);
     cf::populateControlFlowToLLVMConversionPatterns(typeConverter, patterns);
+
+    mlir::populateGpuToLLVMSPVConversionPatterns(typeConverter, patterns);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h:267]: [optional]
```suggestion
    populateGpuToLLVMSPVConversionPatterns(typeConverter, patterns);
```

**Review by whitneywhtsang — APPROVED** [review]

File: test/Conversion/intel/dpas_to_block_layout_convert.mlir [hunk 5]
```diff
@@ -18,7 +18,7 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32
     // CHECK-DAG:           %[[CST_16:.*]] = llvm.mlir.constant(16 : i32) : i32
     // CHECK-DAG:           %[[CST_0:.*]] = llvm.mlir.constant(0 : i32) : i32
     // COM: The following operations is generated for the conversion of DPAS layout to blocked layout.  The conversion replica size is 128*256. So there is 1 round of load/store with synchronization.
-    // CHECK:           %[[threadId_64:.*]] = llvm.call spir_funccc @_Z12get_local_idj(%[[CST_0]]) {function_type = !llvm.func<i64 (i32)>, linkage = #llvm.linkage<external>, passthrough = ["nounwind", "willreturn", ["memory", "0"]], sym_name = "_Z12get_local_idj", visibility_ = 0 : i64} : (i32) -> i64
+    // CHECK:           %[[threadId_64:.*]] = llvm.call spir_funccc @_Z12get_local_idj(%[[CST_0]]) {memory_effects = #llvm.memory_effects<other = none, argMem = none, inaccessibleMem = none>, no_unwind, will_return} : (i32) -> i64
```
Comment (etiotto) [inline test/Conversion/intel/dpas_to_block_layout_convert.mlir:21]: The linkage is gone. Any difference in the generate LLVM IR for the call site (compared to using passthrough attributes) ?

File: third_party/intel/lib/GPUToTritonGEN/GPUToTritonGENPass.cpp [hunk 6]
```diff
@@ -210,22 +210,16 @@ static void populateOpPatterns(LLVMTypeConverter &converter,
...
-      GPUIndexIntrinsicOpLowering<mlir::gpu::BlockDimOp, TritonGEN::BlockDimXOp,
-                                  TritonGEN::BlockDimYOp,
-                                  TritonGEN::BlockDimZOp>,
-      GPUIndexIntrinsicOpLowering<mlir::gpu::GridDimOp, TritonGEN::GridDimXOp,
-                                  TritonGEN::GridDimYOp, TritonGEN::GridDimZOp>,
-      SingleDimLaunchConfigLowering<mlir::gpu::SubgroupIdOp,
-                                    TritonGEN::SubgroupIdOp>,
-      SingleDimLaunchConfigLowering<mlir::gpu::LaneIdOp,
-                                    TritonGEN::SubgroupLocalIdOp>>(converter);
+  // This will ensure that the gpu-to-triton-gen lowering is prefered to the
```
Comment (etiotto) [inline third_party/intel/lib/GPUToTritonGEN/GPUToTritonGENPass.cpp:213]: Add FIXME here to indicate the TritonGEN dialect operations  that should be removed.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp [hunk 7]
```diff
@@ -2172,10 +2172,9 @@ struct IndexCastOpLowering
     if (targetBits == sourceBits)
       return {operands[0][0]};
     if (targetBits < sourceBits)
-      return {rewriter.replaceOpWithNewOp<LLVM::TruncOp>(op, elemTy,
-                                                         operands[0][0])};
-    return {
-        rewriter.replaceOpWithNewOp<LLVM::SExtOp>(op, elemTy, operands[0][0])};
+      return {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp:2175]: Aren't these 2 lines going to gen. the same code as before ?

File: test/Conversion/intel/dpas_to_block_layout_convert.mlir (hunk 5 above)
Comment (FMarno) [reply test/Conversion/intel/dpas_to_block_layout_convert.mlir:21]: I ran 
```
./python/build/cmake.linux-x86_64-cpython-3.10/bin/triton-opt --split-input-file --intel-allocate-shared-memory --convert-triton-intel-gpu-to-llvm --cse -canonicalize ./test/Conversion/intel/dpas_to_block_layout_convert.mlir | ~/compiler/llvm/build/bin/mlir-translate --split-input-file --mlir-to-llvmir
```
before and after the changes, the only differences was many instances of 
```
<   %1555 = call spir_func i64 @_Z12get_local_idj(i32 0) #1
---
>   %1555 = call spir_func i64 @_Z12get_local_idj(i32 0)
```
with
```
attributes #1 = { nounwind willreturn memory(none) }
```

Comment (FMarno) [reply test/Conversion/intel/dpas_to_block_layout_convert.mlir:21]: That's a bit unexpected, since it looks like the passthrough attributes aren't being applied to the callsites, only the function declaration, but I've double checked.

File: third_party/intel/lib/GPUToTritonGEN/GPUToTritonGENPass.cpp (hunk 6 above)
Comment (FMarno) [reply third_party/intel/lib/GPUToTritonGEN/GPUToTritonGENPass.cpp:213]: added a fixme. #2100 should fix that.

File: third_party/intel/CMakeLists.txt (hunk 3 above)
Comment (FMarno) [reply third_party/intel/CMakeLists.txt]: Thanks, I've done that.

File: third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h (hunk 4 above)
Comment (FMarno) [reply third_party/intel/lib/TritonIntelGPUToLLVM/PipelineManager.h]: done :)

**Review by victor-eds — APPROVED** [review]

File: test/Conversion/intel/dpas_to_block_layout_convert.mlir (hunk 5 above)
Comment (etiotto) [reply test/Conversion/intel/dpas_to_block_layout_convert.mlir:21]: That is indeed very unexpected!

**Review by etiotto — APPROVED** [review]
